### PR TITLE
Enabled the LSP Features currently supported for O#

### PR DESCRIPTION
**Pull request by JoeRobich**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3942,100 +3942,133 @@
       {
         "language": "csharp",
         "scopes": {
-          "plainKeyword": [
+          "keyword": [
             "keyword.cs"
           ],
-          "controlKeyword": [
+          "keyword - control": [
```
Comment (JoeRobich) [inline package.json:3948]: Use the Roslyn classifications names as token type.

File: src/features/codeActionProvider.ts [hunk 2]
```diff
@@ -13,7 +13,7 @@ import OptionProvider from '../observers/OptionProvider';
 import { LanguageMiddlewareFeature } from '../omnisharp/LanguageMiddlewareFeature';
 import { buildEditForResponse } from '../omnisharp/fileOperationsResponseEditBuilder';
 
-export default class CodeActionProvider extends AbstractProvider implements vscode.CodeActionProvider<vscode.CodeAction> {
+export default class OmniSharpCodeActionProvider extends AbstractProvider implements vscode.CodeActionProvider<vscode.CodeAction> {
```
Comment (JoeRobich) [inline src/features/codeActionProvider.ts:16]: Make the distinction that these are OmniSharp providers

File: src/features/semanticTokensProvider.ts [hunk 3]
```diff
@@ -254,54 +254,54 @@ tokenTypes[DefaultTokenType.keyword] = SemanticTokenTypes.keyword;
 tokenTypes[DefaultTokenType.number] = SemanticTokenTypes.number;
 tokenTypes[DefaultTokenType.regexp] = SemanticTokenTypes.regexp;
 tokenTypes[DefaultTokenType.operator] = SemanticTokenTypes.operator;
-tokenTypes[DefaultTokenType.namespace] = SemanticTokenTypes.namespace;
+tokenTypes[DefaultTokenType.namespace] = "namespace name";
```
Comment (JoeRobich) [inline src/features/semanticTokensProvider.ts:257]: Use the Roslyn classification names for token types

File: src/omnisharp/engines/IEngine.ts [hunk 4]
```diff
@@ -8,22 +8,36 @@ import { LaunchTarget } from '../launcher';
...
     start(
         cwd: string,
         args: string[],
         launchTarget: LaunchTarget,
         launchInfo: LaunchInfo,
-        options: Options
+        options: Options,
     ): Promise<void>;
     stop(): Promise<void>;
+    registerProviders(
```
Comment (JoeRobich) [inline src/omnisharp/engines/IEngine.ts:27]: Allow the engines to register VS Code providers as needed. LSP won't necessarily need as many.

File: src/omnisharp/engines/StdioEngine.ts [hunk 5]
```diff
@@ -53,6 +80,74 @@ export class StdioEngine implements IEngine {
         );
     }
 
+    async registerProviders(server: OmniSharpServer, optionProvider: OptionProvider, languageMiddlewareFeature: LanguageMiddlewareFeature, eventStream: EventStream, advisor: Advisor, testManager: TestManager): Promise<Disposable> {
+        const documentSelector: vscode.DocumentSelector = {
+            language: 'csharp',
+        };
+
+        const options = optionProvider.GetLatestOptions();
```
Comment (JoeRobich) [inline src/omnisharp/engines/StdioEngine.ts:88]: This is essentially moved as is from the extension.ts

File: src/omnisharp/engines/LspEngine.ts [hunk 6]
```diff
@@ -95,25 +101,24 @@ export class LspEngine implements IEngine {
             middleware: {
                 async provideDefinition(document, position, token, next) {
                     const result = await next(document, position, token);
+                    // Needs Metadata document support
```
Comment (333fred) [inline src/omnisharp/engines/LspEngine.ts:104]: And source generator document support, true for all the locations with this comment.

**Review by 333fred — APPROVED** [review]

**Pull request merged**